diff --git a/libloragw/src/loragw_hal.c b/libloragw/src/loragw_hal.c
index fc03b3f..0499bb9 100644
--- a/libloragw/src/loragw_hal.c
+++ b/libloragw/src/loragw_hal.c
@@ -167,8 +167,7 @@ int lgw_start(void) {
 
     /* Check MCU version (ignore first char of the received version (release/debug) */
     if (strncmp(gw_info.version + 1, mcu_version_string, sizeof mcu_version_string) != 0) {
-        printf("ERROR: MCU version mismatch (expected:%s, got:%s)\n", mcu_version_string, gw_info.version);
-        return -1;
+        printf("WARNING: MCU version mismatch (expected:%s, got:%s)\n", mcu_version_string, gw_info.version + 1);
     }
     printf("INFO: Concentrator MCU version is %s\n", gw_info.version);
 
