diff --git a/lora/rak2246/install.sh b/lora/rak2246/install.sh
index bb896d6..695f9a4 100755
--- a/lora/rak2246/install.sh
+++ b/lora/rak2246/install.sh
@@ -32,7 +32,7 @@ pushd lora_gateway
 
 cp $SCRIPT_DIR/library.cfg ./libloragw/library.cfg
 cp $SCRIPT_DIR/loragw_spi.native.c ./libloragw/src/loragw_spi.native.c
-cp $SCRIPT_DIR/../print_lora_log.sh util_pkt_logger/
+#cp $SCRIPT_DIR/../print_lora_log.sh util_pkt_logger/
 make
 
 popd
diff --git a/lora/rak2246/lora_pkt_fwd.c b/lora/rak2246/lora_pkt_fwd.c
index 1cdf15b..38e478f 100644
--- a/lora/rak2246/lora_pkt_fwd.c
+++ b/lora/rak2246/lora_pkt_fwd.c
@@ -2793,7 +2793,9 @@ void thread_gps(void) {
         if (nb_char <= 0) {
 	    invalid_nb_char_count++;
 	    if ((invalid_nb_char_count % 20000) == 0){
+            #if DEBUG_GPS == 1
                 MSG("WARNING: [gps] read() returned value %d\n", nb_char);
+            #endif
 	    }
             continue;
         }
diff --git a/lora/rak2246/loragw_spi.native.c b/lora/rak2246/loragw_spi.native.c
index a4f4b01..2a1d78f 100644
--- a/lora/rak2246/loragw_spi.native.c
+++ b/lora/rak2246/loragw_spi.native.c
@@ -53,9 +53,8 @@ Maintainer: Sylvain Miermont
 
 #define READ_ACCESS     0x00
 #define WRITE_ACCESS    0x80
-#define SPI_SPEED       2000000
-#define SPI_DEV_PATH    "/dev/spidev0.0"
-//#define SPI_DEV_PATH    "/dev/spidev32766.0"
+#define SPI_SPEED       (getenv("LORAGW_SPI_SPEED")==NULL ? 2000000 : atoi(getenv("LORAGW_SPI_SPEED")))
+#define SPI_DEV_PATH    (getenv("LORAGW_SPI")==NULL ? "/dev/spidev0.0" : getenv("LORAGW_SPI"))
 
 /* -------------------------------------------------------------------------- */
 /* --- PUBLIC FUNCTIONS DEFINITION ------------------------------------------ */
diff --git a/lora/rak2247_spi/install.sh b/lora/rak2247_spi/install.sh
index 8e321db..2acd9ec 100755
--- a/lora/rak2247_spi/install.sh
+++ b/lora/rak2247_spi/install.sh
@@ -31,7 +31,7 @@ fi
 pushd lora_gateway
 
 cp $SCRIPT_DIR/library.cfg ./libloragw/library.cfg
-#cp $SCRIPT_DIR/loragw_spi.native.c ./libloragw/src/loragw_spi.native.c
+cp $SCRIPT_DIR/../rak2246/loragw_spi.native.c ./libloragw/src/loragw_spi.native.c
 make
 
 popd
diff --git a/lora/rak2247_spi/lora_pkt_fwd.c b/lora/rak2247_spi/lora_pkt_fwd.c
index 1cdf15b..38e478f 100644
--- a/lora/rak2247_spi/lora_pkt_fwd.c
+++ b/lora/rak2247_spi/lora_pkt_fwd.c
@@ -2793,7 +2793,9 @@ void thread_gps(void) {
         if (nb_char <= 0) {
 	    invalid_nb_char_count++;
 	    if ((invalid_nb_char_count % 20000) == 0){
+            #if DEBUG_GPS == 1
                 MSG("WARNING: [gps] read() returned value %d\n", nb_char);
+            #endif
 	    }
             continue;
         }
diff --git a/lora/rak2247_usb/install.sh b/lora/rak2247_usb/install.sh
index 91dbbf7..47c3328 100755
--- a/lora/rak2247_usb/install.sh
+++ b/lora/rak2247_usb/install.sh
@@ -23,12 +23,8 @@ apt-get -y install git libftdi-dev libusb-dev
 
 if [ ! -d libmpsse ]; then
     git clone https://github.com/devttys0/libmpsse.git
-	# judge RPiOS64
-	uname -a | grep aarch64
-	if [ $? -eq 0 ] ;then
-		cp -f mpsse.h libmpsse/src
-		cp -f fast.c libmpsse/src
-	fi
+    cp -f mpsse.h libmpsse/src
+    cp -f fast.c libmpsse/src
 fi
 pushd libmpsse/src
 ./configure --disable-python
diff --git a/lora/rak2247_usb/lora_pkt_fwd.c b/lora/rak2247_usb/lora_pkt_fwd.c
index 1cdf15b..38e478f 100644
--- a/lora/rak2247_usb/lora_pkt_fwd.c
+++ b/lora/rak2247_usb/lora_pkt_fwd.c
@@ -2793,7 +2793,9 @@ void thread_gps(void) {
         if (nb_char <= 0) {
 	    invalid_nb_char_count++;
 	    if ((invalid_nb_char_count % 20000) == 0){
+            #if DEBUG_GPS == 1
                 MSG("WARNING: [gps] read() returned value %d\n", nb_char);
+            #endif
 	    }
             continue;
         }
diff --git a/lora/rak7243/lora_pkt_fwd.c b/lora/rak7243/lora_pkt_fwd.c
index ebc8956..c2fb2a4 100644
--- a/lora/rak7243/lora_pkt_fwd.c
+++ b/lora/rak7243/lora_pkt_fwd.c
@@ -2793,7 +2793,9 @@ void thread_gps(void) {
         if (nb_char <= 0) {
 	    invalid_nb_char_count++;
 	    if ((invalid_nb_char_count % 20000) == 0){
+            #if DEBUG_GPS == 1
                 MSG("WARNING: [gps] read() returned value %d\n", nb_char);
+            #endif
 	    }
             continue;
         }
diff --git a/lora/rak7243/loragw_spi.native.c b/lora/rak7243/loragw_spi.native.c
index a4f4b01..2a1d78f 100644
--- a/lora/rak7243/loragw_spi.native.c
+++ b/lora/rak7243/loragw_spi.native.c
@@ -53,9 +53,8 @@ Maintainer: Sylvain Miermont
 
 #define READ_ACCESS     0x00
 #define WRITE_ACCESS    0x80
-#define SPI_SPEED       2000000
-#define SPI_DEV_PATH    "/dev/spidev0.0"
-//#define SPI_DEV_PATH    "/dev/spidev32766.0"
+#define SPI_SPEED       (getenv("LORAGW_SPI_SPEED")==NULL ? 2000000 : atoi(getenv("LORAGW_SPI_SPEED")))
+#define SPI_DEV_PATH    (getenv("LORAGW_SPI")==NULL ? "/dev/spidev0.0" : getenv("LORAGW_SPI"))
 
 /* -------------------------------------------------------------------------- */
 /* --- PUBLIC FUNCTIONS DEFINITION ------------------------------------------ */
diff --git a/lora/rak7243/test_loragw_gps_i2c.c b/lora/rak7243/test_loragw_gps_i2c.c
index d652343..5edff3b 100644
--- a/lora/rak7243/test_loragw_gps_i2c.c
+++ b/lora/rak7243/test_loragw_gps_i2c.c
@@ -198,7 +198,9 @@ int main()
         /* blocking non-canonical read on serial port */
         ssize_t nb_char = read(gps_tty_dev, serial_buff + wr_idx, LGW_GPS_MIN_MSG_SIZE);
         if (nb_char <= 0) {
-            printf("WARNING: [gps] read() returned value %d\n", nb_char);
+            #if DEBUG_GPS == 1
+                printf("WARNING: [gps] read() returned value %d\n", nb_char);
+            #endif
             continue;
         }
         wr_idx += (size_t)nb_char;
diff --git a/lora/rak7243/test_loragw_gps_uart.c b/lora/rak7243/test_loragw_gps_uart.c
index a4164a3..bc625bb 100644
--- a/lora/rak7243/test_loragw_gps_uart.c
+++ b/lora/rak7243/test_loragw_gps_uart.c
@@ -198,7 +198,9 @@ int main()
         /* blocking non-canonical read on serial port */
         ssize_t nb_char = read(gps_tty_dev, serial_buff + wr_idx, LGW_GPS_MIN_MSG_SIZE);
         if (nb_char <= 0) {
-            printf("WARNING: [gps] read() returned value %d\n", nb_char);
+            #if DEBUG_GPS == 1
+                printf("WARNING: [gps] read() returned value %d\n", nb_char);
+            #endif
             continue;
         }
         wr_idx += (size_t)nb_char;
